### ISSUE TITLE
[BUG]  demo data failures

### ISSUE THREAD

**Issue by dkp** [cpp-lln-lab/bidsMReye#243]
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Expected Behavior

I expected the demo data to work as described here: https://bidsmreye.readthedocs.io/en/latest/demo.html

### Current Behavior

**Problem 1: Fixed**: The Docker command did not run because it needed the --action flag.  Fixed by adding the --action flag (but the documentation should be updated)

**Problem 2: Fixed**: The Docker container had permissions trouble writing to the derivatives directory it had created.
Fixed: By as apptainer instead (but you might consider providing more information about the permission problems.

**Success (I think)**: I ran `--action prepare` and got:

```bash
[dema:bidsmreye_demo () 2086] $ ./bidsmreye_demo.sh
2024-12-21 13:31:50.334842: I tensorflow/core/platform/cpu_feature_guard.cc:193] This TensorFlow binary is optimized with oneAPI Deep Neural Network Library (oneDNN) to use the following CPU instructions in performance-critical operations:  AVX2 FMA
To enable them in other operations, rebuild TensorFlow with the appropriate compiler flags.
2024-12-21 13:31:50.680408: W tensorflow/compiler/xla/stream_executor/platform/default/dso_loader.cc:64] Could not load dynamic library 'libcudart.so.11.0'; dlerror: libcudart.so.11.0: cannot open shared object file: No such file or directory; LD_LIBRARY_PATH: /.singularity.d/libs
2024-12-21 13:31:50.680433: I tensorflow/compiler/xla/stream_executor/cuda/cudart_stub.cc:29] Ignore above cudart dlerror if you do not have a GPU set up on your machine.
2024-12-21 13:31:51.692280: W tensorflow/compiler/xla/stream_executor/platform/default/dso_loader.cc:64] Could not load dynamic library 'libnvinfer.so.7'; dlerror: libnvinfer.so.7: cannot open shared object file: No such file or directory; LD_LIBRARY_PATH: /.singularity.d/libs
2024-12-21 13:31:51.692457: W tensorflow/compiler/xla/stream_executor/platform/default/dso_loader.cc:64] Could not load dynamic library 'libnvinfer_plugin.so.7'; dlerror: libnvinfer_plugin.so.7: cannot open shared object file: No such file or directory; LD_LIBRARY_PATH: /.singularity.d/libs
2024-12-21 13:31:51.692472: W tensorflow/compiler/tf2tensorrt/utils/py_utils.cc:38] TF-TRT Warning: Cannot dlopen some TensorRT libraries. If you would like to use Nvidia GPU with TensorRT, please make sure the missing libraries mentioned above are installed properly.
WARNING
[13:31:55] WARNING  bidsMReye - [13:31:55] - log level: WARNING                                                                               bidsmreye.py:112
Mask 0/2, Sum: 4.251, Mean 0.354, Std 0.462, Median 0.012
Mask 1/2, Sum: 11.329, Mean 0.944, Std 1.870, Median 0.088
Mask 2/2, Sum: 1.140, Mean 0.095, Std 0.878, Median 0.001
/opt/miniconda-latest/envs/bidsmreye/lib/python3.9/site-packages/deepmreye/preprocess.py:340: RuntimeWarning:

Mean of empty slice

/opt/miniconda-latest/envs/bidsmreye/lib/python3.9/site-packages/numpy/lib/nanfunctions.py:1872: RuntimeWarning:

Degrees of freedom <= 0 for slice.

Saving eye data (84, 47, 29, 18) and targets (84, 10, 2) (NaN 0) to file /home/neuro/outputs/bidsmreye/sub-01_task-auditory_space-MNI152NLin6Asym_desc-nolabel_bidsmreye.npz
Mask 0/2, Sum: 12.325, Mean 1.027, Std 4.988, Median 0.023
Mask 1/2, Sum: 15.340, Mean 1.278, Std 2.608, Median 0.052
Mask 2/2, Sum: -0.028, Mean -0.002, Std 0.833, Median 0.010
Saving eye data (84, 47, 29, 18) and targets (84, 10, 2) (NaN 0) to file /home/neuro/outputs/bidsmreye/sub-01_task-auditory_space-T1w_desc-nolabel_bidsmreye.npz
``` 

**Problem 3** `--action generalize` (see error message below)

### Error message

[dema:bidsmreye_demo () 2086] $ ./bidsmreye_demo.sh
2024-12-21 13:31:50.334842: I tensorflow/core/platform/cpu_feature_guard.cc:193] This TensorFlow binary is optimized with oneAPI Deep Neural Network Library (oneDNN) to use the following CPU instructions in performance-critical operations:  AVX2 FMA
To enable them in other operations, rebuild TensorFlow with the appropriate compiler flags.
2024-12-21 13:31:50.680408: W tensorflow/compiler/xla/stream_executor/platform/default/dso_loader.cc:64] Could not load dynamic library 'libcudart.so.11.0'; dlerror: libcudart.so.11.0: cannot open shared object file: No such file or directory; LD_LIBRARY_PATH: /.singularity.d/libs
2024-12-21 13:31:50.680433: I tensorflow/compiler/xla/stream_executor/cuda/cudart_stub.cc:29] Ignore above cudart dlerror if you do not have a GPU set up on your machine.
2024-12-21 13:31:51.692280: W tensorflow/compiler/xla/stream_executor/platform/default/dso_loader.cc:64] Could not load dynamic library 'libnvinfer.so.7'; dlerror: libnvinfer.so.7: cannot open shared object file: No such file or directory; LD_LIBRARY_PATH: /.singularity.d/libs
2024-12-21 13:31:51.692457: W tensorflow/compiler/xla/stream_executor/platform/default/dso_loader.cc:64] Could not load dynamic library 'libnvinfer_plugin.so.7'; dlerror: libnvinfer_plugin.so.7: cannot open shared object file: No such file or directory; LD_LIBRARY_PATH: /.singularity.d/libs
2024-12-21 13:31:51.692472: W tensorflow/compiler/tf2tensorrt/utils/py_utils.cc:38] TF-TRT Warning: Cannot dlopen some TensorRT libraries. If you would like to use Nvidia GPU with TensorRT, please make sure the missing libraries mentioned above are installed properly.
WARNING
[13:31:55] WARNING  bidsMReye - [13:31:55] - log level: WARNING                                                                               bidsmreye.py:112
Mask 0/2, Sum: 4.251, Mean 0.354, Std 0.462, Median 0.012
Mask 1/2, Sum: 11.329, Mean 0.944, Std 1.870, Median 0.088
Mask 2/2, Sum: 1.140, Mean 0.095, Std 0.878, Median 0.001
/opt/miniconda-latest/envs/bidsmreye/lib/python3.9/site-packages/deepmreye/preprocess.py:340: RuntimeWarning:

Mean of empty slice

/opt/miniconda-latest/envs/bidsmreye/lib/python3.9/site-packages/numpy/lib/nanfunctions.py:1872: RuntimeWarning:

Degrees of freedom <= 0 for slice.

Saving eye data (84, 47, 29, 18) and targets (84, 10, 2) (NaN 0) to file /home/neuro/outputs/bidsmreye/sub-01_task-auditory_space-MNI152NLin6Asym_desc-nolabel_bidsmreye.npz
Mask 0/2, Sum: 12.325, Mean 1.027, Std 4.988, Median 0.023
Mask 1/2, Sum: 15.340, Mean 1.278, Std 2.608, Median 0.052
Mask 2/2, Sum: -0.028, Mean -0.002, Std 0.833, Median 0.010
Saving eye data (84, 47, 29, 18) and targets (84, 10, 2) (NaN 0) to file /home/neuro/outputs/bidsmreye/sub-01_task-auditory_space-T1w_desc-nolabel_bidsmreye.npz
[dema:bidsmreye_demo () 2087] $ ./bidsmreye_demo.sh
2024-12-21 13:36:27.766460: I tensorflow/core/platform/cpu_feature_guard.cc:193] This TensorFlow binary is optimized with oneAPI Deep Neural Network Library (oneDNN) to use the following CPU instructions in performance-critical operations:  AVX2 FMA
To enable them in other operations, rebuild TensorFlow with the appropriate compiler flags.
2024-12-21 13:36:28.115113: W tensorflow/compiler/xla/stream_executor/platform/default/dso_loader.cc:64] Could not load dynamic library 'libcudart.so.11.0'; dlerror: libcudart.so.11.0: cannot open shared object file: No such file or directory; LD_LIBRARY_PATH: /.singularity.d/libs
2024-12-21 13:36:28.115140: I tensorflow/compiler/xla/stream_executor/cuda/cudart_stub.cc:29] Ignore above cudart dlerror if you do not have a GPU set up on your machine.
2024-12-21 13:36:29.149578: W tensorflow/compiler/xla/stream_executor/platform/default/dso_loader.cc:64] Could not load dynamic library 'libnvinfer.so.7'; dlerror: libnvinfer.so.7: cannot open shared object file: No such file or directory; LD_LIBRARY_PATH: /.singularity.d/libs
2024-12-21 13:36:29.149773: W tensorflow/compiler/xla/stream_executor/platform/default/dso_loader.cc:64] Could not load dynamic library 'libnvinfer_plugin.so.7'; dlerror: libnvinfer_plugin.so.7: cannot open shared object file: No such file or directory; LD_LIBRARY_PATH: /.singularity.d/libs
2024-12-21 13:36:29.149794: W tensorflow/compiler/tf2tensorrt/utils/py_utils.cc:38] TF-TRT Warning: Cannot dlopen some TensorRT libraries. If you would like to use Nvidia GPU with TensorRT, please make sure the missing libraries mentioned above are installed properly.
WARNING
[13:36:33] WARNING  bidsMReye - [13:36:33] - log level: WARNING                                                                               bidsmreye.py:112

Training set (/home/neuro/outputs/bidsmreye/sub-01/func) contains 1 subjects:
['sub-01_task-auditory_space-MNI152NLin6Asym_desc-nolabel_bidsmreye']
Test set (/home/neuro/outputs/bidsmreye/sub-01/func) contains 1 subjects:
['sub-01_task-auditory_space-MNI152NLin6Asym_desc-nolabel_bidsmreye']

2024-12-21 13:36:33.811777: W tensorflow/compiler/xla/stream_executor/platform/default/dso_loader.cc:64] Could not load dynamic library 'libcuda.so.1'; dlerror: libcuda.so.1: cannot open shared object file: No such file or directory; LD_LIBRARY_PATH: /.singularity.d/libs
2024-12-21 13:36:33.811811: W tensorflow/compiler/xla/stream_executor/cuda/cuda_driver.cc:265] failed call to cuInit: UNKNOWN ERROR (303)
2024-12-21 13:36:33.811828: I tensorflow/compiler/xla/stream_executor/cuda/cuda_diagnostics.cc:156] kernel driver does not appear to be running on this host (dema): /proc/driver/nvidia/version does not exist
2024-12-21 13:36:33.812459: I tensorflow/core/platform/cpu_feature_guard.cc:193] This TensorFlow binary is optimized with oneAPI Deep Neural Network Library (oneDNN) to use the following CPU instructions in performance-critical operations:  AVX2 FMA
To enable them in other operations, rebuild TensorFlow with the appropriate compiler flags.
/opt/miniconda-latest/envs/bidsmreye/lib/python3.9/site-packages/numpy/lib/function_base.py:2854: RuntimeWarning:

invalid value encountered in divide

/opt/miniconda-latest/envs/bidsmreye/lib/python3.9/site-packages/numpy/lib/function_base.py:2855: RuntimeWarning:

invalid value encountered in divide

/opt/miniconda-latest/envs/bidsmreye/lib/python3.9/site-packages/numpy/lib/function_base.py:2854: RuntimeWarning:

invalid value encountered in divide

/opt/miniconda-latest/envs/bidsmreye/lib/python3.9/site-packages/numpy/lib/function_base.py:2855: RuntimeWarning:

invalid value encountered in divide

/opt/miniconda-latest/envs/bidsmreye/lib/python3.9/site-packages/numpy/lib/function_base.py:2854: RuntimeWarning:

invalid value encountered in divide

### Environment

```markdown
- OS:Ubuntu 24.04
- Tried Docker
- Tried Apptainer
```

### Branch

```markdown
[dema:bidsmreye_demo () 2087] $ apptainer run ${SIF}/bidsmreye.sif --version
2024-12-21 14:03:59.370321: I tensorflow/core/platform/cpu_feature_guard.cc:193] This TensorFlow binary is optimized with oneAPI Deep Neural Network Library (oneDNN) to use the following CPU instructions in performance-critical operations:  AVX2 FMA
To enable them in other operations, rebuild TensorFlow with the appropriate compiler flags.
2024-12-21 14:03:59.719681: W tensorflow/compiler/xla/stream_executor/platform/default/dso_loader.cc:64] Could not load dynamic library 'libcudart.so.11.0'; dlerror: libcudart.so.11.0: cannot open shared object file: No such file or directory; LD_LIBRARY_PATH: /.singularity.d/libs
2024-12-21 14:03:59.719705: I tensorflow/compiler/xla/stream_executor/cuda/cudart_stub.cc:29] Ignore above cudart dlerror if you do not have a GPU set up on your machine.
2024-12-21 14:04:00.741471: W tensorflow/compiler/xla/stream_executor/platform/default/dso_loader.cc:64] Could not load dynamic library 'libnvinfer.so.7'; dlerror: libnvinfer.so.7: cannot open shared object file: No such file or directory; LD_LIBRARY_PATH: /.singularity.d/libs
2024-12-21 14:04:00.741659: W tensorflow/compiler/xla/stream_executor/platform/default/dso_loader.cc:64] Could not load dynamic library 'libnvinfer_plugin.so.7'; dlerror: libnvinfer_plugin.so.7: cannot open shared object file: No such file or directory; LD_LIBRARY_PATH: /.singularity.d/libs
2024-12-21 14:04:00.741674: W tensorflow/compiler/tf2tensorrt/utils/py_utils.cc:38] TF-TRT Warning: Cannot dlopen some TensorRT libraries. If you would like to use Nvidia GPU with TensorRT, please make sure the missing libraries mentioned above are installed properly.
bidsMReye version 0.3.0
```

### Anything else?

Apptainer Script contents:

# Just to be safe, create the derivatives subdirectory for this app, if it does not exist.
if [ ! -d $PWD/moae_fmriprep/derivatives ]; then
    mkdir -p $PWD/moae_fmriprep/derivatives
fi

export APP=${SIF}/bidsmreye.sif

## lots of permission issues

# docker run --rm -i -v $PWD/moae_fmriprep:/home/neuro/data  -v $PWD/moae_fmriprep/derivatives:/home/neuro/outputs/ cpplab/bidsmreye:latest /home/neuro/data/ /home/neuro/outputs/ participant --action prepare

## Success! no permission complaints and it seems to only warn about the cuda libraries
# apptainer run --cleanenv --bind $PWD/moae_fmriprep:/home/neuro/data --bind $PWD/moae_fmriprep/derivatives:/home/neuro/outputs/ ${APP} /home/neuro/data /home/neuro/outputs participant --action prepare

## Fail: The generalize step failswith an "invalid value encountered in divide" error
apptainer run --cleanenv --bind $PWD/moae_fmriprep:/home/neuro/data --bind $PWD/moae_fmriprep/derivatives:/home/neuro/outputs/ ${APP} /home/neuro/data /home/neuro/outputs participant --action generalize
